crates/cubecl-opt/src/lib.rs

```diff
@@ -0,0 +1,359 @@
+            }
+        }
+    }
+
+    fn node_ids(&self) -> Vec<NodeIndex> {
+        self.program.node_indices().collect()
+    }
+
+    fn ssa_transform(&mut self) {
+        self.program.fill_dom_frontiers();
+        self.program.place_phi_nodes();
+        self.version_program();
+        self.program.variables.clear();
+        for block in self.node_ids() {
+            self.program[block].writes.clear();
+        }
+    }
+
+    pub fn current_block_mut(&mut self) -> &mut BasicBlock {
+        &mut self.program[self.current_block.unwrap()]
+    }
+
+    pub fn current_block(&self) -> &BasicBlock {
+        &self.program[self.current_block.unwrap()]
+    }
+
+    pub fn predecessors(&self, block: NodeIndex) -> Vec<NodeIndex> {
+        self.program
+            .edges_directed(block, Direction::Incoming)
+            .map(|it| it.source())
+            .collect()
+    }
+
+    pub fn sucessors(&self, block: NodeIndex) -> Vec<NodeIndex> {
+        self.program
+            .edges_directed(block, Direction::Outgoing)
+            .map(|it| it.target())
+            .collect()
+    }
+
+    pub fn block(&self, block: NodeIndex) -> &BasicBlock {
+        &self.program[block]
+    }
+
+    pub fn parse_scope(&mut self, mut scope: Scope) {
+        let processed = scope.process();
+
+        for var in processed.variables {
+            if let Variable::Local { id, item, depth } = var {
+                self.program.variables.insert((id, depth), item);
+            }
+        }
+
+        for instruction in processed.operations {
+            match instruction {
+                Operation::Branch(branch) => self.parse_control_flow(branch),
+                Operation::Procedure(proc) => self.compile_procedure(proc, scope.clone()),
+                Operation::Operator(Operator::Slice(slice_op)) => {
+                    let out_id = match &slice_op.out {
+                        Variable::Slice { id, depth, .. } => (*id, *depth),
+                        _ => unreachable!(),
+                    };
+                    self.program.slices.insert(
+                        out_id,
+                        Slice {
+                            start: slice_op.start,
+                            end: slice_op.end,
+                            end_op: None,
+                            const_len: None,
+                        },
+                    );
+                    let mut op = Operation::Operator(Operator::Slice(slice_op));
+                    self.visit_operation(&mut op, |_, _| {}, |opt, var| opt.write_var(var));
+                    self.current_block_mut().ops.borrow_mut().push(op);
+                }
+                mut other => {
+                    self.visit_operation(&mut other, |_, _| {}, |opt, var| opt.write_var(var));
+                    self.current_block_mut().ops.borrow_mut().push(other);
+                }
+            }
+        }
+    }
+
+    fn compile_procedure(&mut self, proc: Procedure, mut scope: Scope) {
```

Comment:
If the optimizer also responsible for recompiling the scope?

Reply:
Yes, the optimizer completely takes over compiling of the scope. Crates using the optimizer should only need to work off the optimizer's `BasicBlock`s. This means it also needs to push the procedure instructions.